src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsReadStream.cs

```diff
@@ -0,0 +1,355 @@
+        /// </summary>
+        /// <param name="underlyingStream">The underlying stream to read from</param>
+        public TdsReadStream(Stream underlyingStream)
+        {
+            _readBuffer = new byte[TdsEnums.DEFAULT_LOGIN_PACKET_SIZE];
+            _underlyingStream = underlyingStream;
+        }
+
+        #endregion
+
+        #region Public Properties
+        /// <inheritdoc />
+        public override bool CanRead => _underlyingStream != null;
+
+        /// <inheritdoc />
+        public override bool CanSeek => false;
+
+        /// <inheritdoc />
+        public override bool CanWrite => false;
+
+        /// <inheritdoc />
+        public override long Length => throw new NotSupportedException();
+
+
+        /// <inheritdoc />
+        public virtual byte ReadPacketHeaderType { get; private set; }
+
+        /// <inheritdoc />
+        public virtual byte ReadPacketStatus { get; private set; }
+
+        /// <inheritdoc />
+        public override long Position 
+        {
+            get => throw new NotSupportedException(); 
+            set => throw new NotSupportedException(); 
+        }
+
+        /// <inheritdoc />
+        public virtual int Spid { get; private set; }
+
+        #endregion
+
+        #region Protected Methods
+
+        /// <inheritdoc />
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                _underlyingStream.Dispose();
+                _underlyingStream = null;
+                _readBuffer = null;
+            }
+            base.Dispose(disposing);
+        }
+
+        #endregion
+
+        #region Public Methods
+
+
+        /// <inheritdoc />
+        public override async ValueTask DisposeAsync()
+        {
+            await _underlyingStream.DisposeAsync();
+            _underlyingStream = null;
+            _readBuffer = null;
+        }
+
+        /// <inheritdoc />
+        public override void Flush() => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public async ValueTask<byte> PeekByteAsync(bool isAsync, CancellationToken ct)
+        {
+            // If we have logically finished reading the packet, or if we have 
+            // reached the end of the buffer, then we need to position the buffer at the beginning of next
+            // packet start.
+            await PrepareBufferAsync(isAsync, ct).ConfigureAwait(false);
+            return _readBuffer[_readIndex];
+        }
+
+        /// <inheritdoc />
+        public override int Read(Span<byte> buffer)
+        {
+            int lengthToFill = buffer.Length;
+            int totalRead = 0;
+            while (lengthToFill > 0)
+            {
+                PrepareBufferAsync(isAsync: false, CancellationToken.None).ConfigureAwait(false);
+
+                int lengthToCopy = MinDataAvailableBeforeRead(lengthToFill);
+                ReadOnlySpan<byte> copyFrom = _readBuffer.AsSpan(_readIndex, lengthToCopy);
```

Comment:
Suggestion: You could also consider reading via [MemoryStream](https://learn.microsoft.com/en-us/dotnet/api/system.io.memorystream?view=net-8.0) instead of readBuffer, and read as:
`int bytesRead = _memoryStream.Read(buffer.Slice(totalRead, lengthToCopy));`

Reply:
@cheenamalhotra I didn't completely understand the comment about memory stream, and in what cases would it help?